Add API Documentation for Multi-Team Tokens (#875)
hashicorp/terraform-docs-common: website/docs/cloud-docs/api-docs/team-tokens.mdx

> ---

Author: Rather than creating a brand new page for the new API, I decided to break the existing team token API documentation into two sections: "Team tokens API reference" and "Legacy team tokens API reference." Since both APIs will exist simultaneously, I didn't want to make the API tokens site navigation confusing, break existing references to this documentation, or create confusion for users who regularly use these docs. Furthermore, both the old and new API share a lot of the same background information and keeping the two coupled and reusing some of that information felt like the more straightforward thing to do. This all being said, though, I don't feel strongly about coupling or decoupling the two APIs and would love to hear what other folks think!

# Team tokens API reference

> Team API tokens grant access to a team's workspaces. Teams are not limited to a single token, and can have multiple, valid tokens at a time. Team tokens are not are not associated with a specific user.

Reviewer: ```suggestion
Team API tokens grant access to a team's workspaces. Teams are not limited to a single token, and can have multiple tokens at a time. Team tokens are not are not associated with a specific user.
```

> Teams relying on the [**legacy**](/terraform/cloud-docs/api-docs/team-tokens#legacy-team-tokens-api-reference) team token API (`/teams/authentication-token`), can only have a **single**, valid token at a time. Generating a new legacy token revokes any previously existing tokens, replacing it with the new team token.

Reviewer: ```suggestion
Teams relying on the [**legacy**](/terraform/cloud-docs/api-docs/team-tokens#legacy-team-tokens-api-reference) team token API (`/teams/authentication-token`), can only have a **single**, valid token at a time. Generating a new legacy token revokes any previously existing tokens, replacing it with the new team token.
```

Does "revokes any previously existing tokens" include tokens under the `/teams/:team_id/authentication-tokens` endpoint, or just the legacy endpoint?

Reviewer 2: ```suggestion
Teams relying on the [**legacy**](/terraform/cloud-docs/api-docs/team-tokens#legacy-team-tokens-api-reference) team token API (`/teams/authentication-token`), can only create a **single**, valid token at a time. Generating a new legacy token when one already exists for the team revokes the existing legacy token, replacing it with a new team token.
```
Maybe something like this to clarify how the legacy API works?

Author: Sure! I've updated this.

You can create and delete team tokens and list an organization's team tokens.

## Generate a new team token

Generates a new team token.

| Method | Path                                 |
| :----- | :----------------------------------- |
| POST   | /teams/:team_id/authentication-tokens |

> This endpoint returns the secret text of the new authentication token. You can only access this token when you create it and can not recover it later.

Reviewer: ```suggestion
This endpoint returns the secret text of the new authentication token. You can only access the secret when you create it and cannot recover it later.
```

Author: Great catch! 👀

| [404][] | [JSON API error object][]                               | Team Token not found, or unauthorized to view the Team Token |

### Sample request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request GET \
  https://app.terraform.io/api/v2/authentication-tokens/at-6yEmxNAhaoQLH1Da
```

### Sample response

```json
{
  "data": {
    "id": "at-6yEmxNAhaoQLH1Da",
    "type": "authentication-tokens",
    "attributes": {
      "created-at": "2017-11-29T19:18:09.976Z",
      "last-used-at": null,
      "description": "Team API token for team ABC",
      "token": "QnbSxjjhVMHJgw.atlasv1.gxZnWIjI5j752DGqdwEUVLOFf0mtyaQ00H9bA1j90qWb254lEkQyOdfqqcq9zZL7Sm0",
      "expired-at": "2023-04-06T12:00:00.000Z"
    },
    "relationships": {
      "team": {
        "data": {
          "id": "team-LnREdjodkvZFGdXL",
          "type": "teams"
        }
      },
      "created-by": {
        "data": {
          "id": "user-MA4GL63FmYRpSFxa",
          "type": "users"
        }
      }
    }
  }
}
```

# Legacy team tokens API reference

> 

Reviewer: Maybe a bit of introductory text? How about something like:

```
Legacy team API tokens grant access to a team's workspaces. Each team can have an API token that is not associated with a specific user. You can create and delete team tokens and list an organization's team tokens. The [team tokens API](/terraform/cloud-docs/api-docs/team-tokens#team-tokens-api-reference) includes the same functionality as legacy team tokens, and allows you to provision multiple tokens per team.
```

## Generate a new team token

Generates a new team token and overrides existing token if one exists.

| Method | Path                                 |
| :----- | :----------------------------------- |

> | POST   | /teams/:team_id/authentication-tokens |

Reviewer: Is the legacy API endpoint changing? This would break existing scripts, wouldn't it?

Author: The legacy endpoints are not changing, and existing scripts using the legacy endpoint should not break! This looks like a typo and I've updated it - great catch 👀

You can create and delete team tokens and list an organization's team tokens.

## Generate a new team token

Generates a new team token.

| Method | Path                                 |
| :----- | :----------------------------------- |
| POST   | /teams/:team_id/authentication-tokens |

This endpoint returns the secret text of the new authentication token. You can only access this token when you create it and can not recover it later.

### Parameters

- `:team_id` (`string: <required>`) - specifies the team ID for generating the team token

### Request body

This POST endpoint requires a JSON object with the following properties as a request payload.

| Key path                      | Type   | Default | Description                                                                                                     |
| ----------------------------- | ------ | ------- | --------------------------------------------------------------------------------------------------------------- |
| `data.type`                   | string |         | Must be `"authentication-token"`.                                                                               |

> | `data.attributes.description` | string |         | The description of the team token. Each description **must** be unique.                                         |

Reviewer 3: I think we should mention that description is unique within a context of a team.

Author: It's mentioned on this line!

Reviewer 3: Sorry, I meant the user might not know if it should be unique within the whole organization or unique within a team. Or is it obvious enough?

    "attributes": {
      "description": "Team API token for team ABC",
      "expired-at": "2023-04-06T12:00:00.000Z"
    }
  }
}
```

### Sample request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  --data @payload.json \
  https://app.terraform.io/api/v2/teams/team-BUHBEM97xboT8TVz/authentication-tokens
```

### Sample response

```json
{
  "data": {
    "id": "4111797",
    "type": "authentication-tokens",
    "attributes": {
      "created-at": "2017-11-29T19:18:09.976Z",
      "last-used-at": null,
      "description": "Team API token for team ABC",
      "token": "QnbSxjjhVMHJgw.atlasv1.gxZnWIjI5j752DGqdwEUVLOFf0mtyaQ00H9bA1j90qWb254lEkQyOdfqqcq9zZL7Sm0",
      "expired-at": "2023-04-06T12:00:00.000Z"
    },
    "relationships": {
      "team": {
        "data": {
          "id": "team-Y7RyjccPVBKVEdp7",
          "type": "teams"
        }
      },
      "created-by": {
        "data": {
          "id": "user-62goNpx1ThQf689e",
          "type": "users"
        }
      }
    }
  }
}

```

## Delete the team token

| Method | Path                                 |

> | :----- | :----------------------------------- |

Reviewer 3: The RFC mentions an api like this: /authentication-tokens/:id.

# Team tokens API reference

Team API tokens grant access to a team's workspaces. Teams are not limited to a single token, and can have multiple tokens at a time. Team tokens are not associated with a specific user.

> Teams relying on the [**legacy**](/terraform/cloud-docs/api-docs/team-tokens#legacy-team-tokens-api-reference) team token API (`/teams/authentication-token`), can only create a **single**, valid token at a time. Generating a new legacy token when one already exists for the team revokes the existing legacy token, replacing it with a new team token.

Reviewer 4: Should there be a team ID param in the legacy API path?

Author: Yup! Great catch. I've updated this ✔️

You can create and delete team tokens and list an organization's team tokens.

## Generate a new team token

Generates a new team token.

| Method | Path                                 |
| :----- | :----------------------------------- |
| POST   | /teams/:team_id/authentication-tokens |

This endpoint returns the secret text of the new authentication token. You can only access the secret when you create it and cannot recover it later.

### Parameters

- `:team_id` (`string: <required>`) - specifies the team ID for generating the team token

### Request body

This POST endpoint requires a JSON object with the following properties as a request payload.

| Key path                      | Type   | Default | Description                                                                                                     |
| ----------------------------- | ------ | ------- | --------------------------------------------------------------------------------------------------------------- |

> | `data.type`                   | string |         | Must be `"authentication-tokens"`.                                                                               |

Author: Updated this to be the plural `authentication-tokens` per [this](https://hashicorp.slack.com/archives/C08FHA29X4Z/p1745425722676109?thread_ts=1745355247.136919&cid=C08FHA29X4Z) Slack comment!

| `data.attributes.description` | string |         | The description of the team token. Each description **must** be unique within the context of the team.          |
| `data.attributes.expired-at`  | string | `null`  | The UTC date and time that the Team Token will expire, in ISO 8601 format. If omitted or set to `null` the token will never expire. |

### Sample payload

```json
{
  "data": {

>     "type": "authentication-tokens",

Author: Updated this to be the plural `authentication-tokens` per [this](https://hashicorp.slack.com/archives/C08FHA29X4Z/p1745425722676109?thread_ts=1745355247.136919&cid=C08FHA29X4Z) Slack comment!

| [404][] | [JSON API error object][]                               | Team not found.                                                                       |

### Query parameters

This endpoint supports pagination [with standard URL query parameters](/terraform/cloud-docs/api-docs#query-parameters) and searching with the `q` parameter. Remember to percent-encode `[` as `%5B` and `]` as `%5D` if your tooling doesn't automatically encode URLs.

| Parameter      | Description                                                         |
|----------------|---------------------------------------------------------------------|
| `page[number]` | **Optional.** If omitted, the endpoint returns the first page.       |
| `page[size]`   | **Optional.** If omitted, the endpoint returns 20 tokens per page.  |
| `q`            | **Optional.** A search query string. You can search for a team authentication token using the team name. |
| `sort`         | **Optional.** Allows sorting the team tokens by `"created-by"`, `"expired-at"`, and `"last-used-at"`. Prepending a hyphen to the sort parameter reverses the order. If omitted, the default sort order ascending. |

### Sample response

> ```json

Author: This is based off of the existing documentation, which did not include a sample request! According to the docs, it looks like users can query using a team name:
> A search query string. You can search for a team authentication token using the team name.

            "type": "teams"
          }
        },
        "created-by": {
          "data": {
            "id": "user-ccU6h629sszLJBpY",
            "type": "users"
          }
        }
      }
    },
    {
      "id": "at-qfc2wqqJ1T5sCamM",
      "type": "authentication-tokens",
      "attributes": {
        "created-at": "2024-06-19T18:44:44.051Z",
        "last-used-at": null,
        "description": "Team API token for team XYZ",
        "token": null,
        "expired-at": "2024-07-19T18:44:43.818Z"
      },
      "relationships": {
        "team": {
          "data": {
            "id": "team-58pFiBffTLMxLphR",
            "type": "teams"
          }
        },
        "created-by": {
          "data": {
            "id": "user-ccU6h629hhzLJBpY",
            "type": "users"
          }
        }
      }
    },
  ]
}
```

## Show a team token

Use this endpoint to display a particular [team token](/terraform/cloud-docs/users-teams-organizations/teams#api-tokens).

`GET /authentication-tokens/:token_id`

| Parameter   | Description               |
| ----------- | ------------------------- |
| `:token_id` | The ID of the Team Token. |

> The object returned by this endpoint only contains metadata, and does not include the secret text of the authentication token. A token's secret test is only shown upon creation, and cannot be recovered later.

Author: Great catch! 😄

| [404][] | [JSON API error object][]                               | Team Token not found, or unauthorized to view the Team Token |

### Sample request

```shell
curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request GET \
  https://app.terraform.io/api/v2/authentication-tokens/at-6yEmxNAhaoQLH1Da
```

### Sample response

```json
{
  "data": {
    "id": "at-6yEmxNAhaoQLH1Da",
    "type": "authentication-tokens",
    "attributes": {
      "created-at": "2017-11-29T19:18:09.976Z",
      "last-used-at": null,
      "description": "Team API token for team ABC",
      "token": "QnbSxjjhVMHJgw.atlasv1.gxZnWIjI5j752DGqdwEUVLOFf0mtyaQ00H9bA1j90qWb254lEkQyOdfqqcq9zZL7Sm0",
      "expired-at": "2023-04-06T12:00:00.000Z"
    },
    "relationships": {
      "team": {
        "data": {
          "id": "team-LnREdjodkvZFGdXL",
          "type": "teams"
        }
      },
      "created-by": {
        "data": {
          "id": "user-MA4GL63FmYRpSFxa",
          "type": "users"
        }
      }
    }
  }
}
```

> # Legacy team tokens API reference

Reviewer 2: I'm wondering if we should remove the `List team tokens` section under the legacy team tokens because it's the same API as above. 🤔

Legacy team API tokens grant access to a team's workspaces. Each team can have an API token that is not associated with a specific user.
You can create and delete team tokens and list an organization's team tokens.

> The [team tokens API](/terraform/cloud-docs/api-docs/team-tokens) includes the same functionality as legacy team tokens, and allows you to provision multiple tokens per team.

Reviewer 2: Should we add some clarification here that these legacy APIs apply to only the token that do not have a description?

Reviewer 5: Legacy tokens with the new API can set a description fyi

| :----- | :----------------------------------- |

> | POST   | /teams/:team_id/authentication-token |

Reviewer 5: If the sample payload below ends in `s` then should this also, @mkam  ?

<img width="950" alt="image" src="https://github.com/user-attachments/assets/edc4f41d-d675-4443-b2e1-f0f74483ee3f" />

Author: Please correct me if I'm misunderstanding, but I believe that the comment that you're referring to meant to say `authentication-token`, singular, rather than `authentication-tokens`, plural. This being said, I've updated @mkam's comment and will leave this as-is since it's the legacy API!

> This endpoint returns the secret text of the new authentication token. You can only access the secret when you create it and cannot recover it later.

Reviewer: ```suggestion
This endpoint returns the secret text of the new authentication token. You can only access the secret text when you create it and cannot recover it later.
```

| Key path                      | Type   | Default | Description                                                                                                     |
| ----------------------------- | ------ | ------- | --------------------------------------------------------------------------------------------------------------- |
| `data.type`                   | string |         | Must be `"authentication-token"`.                                                                               |

> | `data.attributes.description` | string | `null`  | The description of the team token, if supplied, otherwise the description will be `null`. The description **must** be unique within the context of the team, meaning legacy tokens may only have one `null` or one descriptive description at a time. |

Reviewer: So this change includes adding a description field to the legacy team token API? Just confirming, as line 301 above implies that description is a new-API-only feature.

  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  --data @payload.json \

>   https://app.terraform.io/api/v2/teams/team-BUHBEM97xboT8TVz/authentication-tokens

Reviewer 2: This should be `authentication-tokens` here since it's for the legacy APIs.

Author: Great catch!
